File: commons/javafx/src/main/java/com/cedarsoft/commons/javafx/ProgressIndicatorBar.java
package com.cedarsoft.commons.javafx;

import javax.annotation.Nonnull;

import javafx.beans.property.DoubleProperty;
import javafx.beans.property.SimpleDoubleProperty;
import javafx.scene.control.ProgressBar;
import javafx.scene.control.ProgressIndicator;
import javafx.scene.layout.StackPane;
import javafx.scene.paint.Color;
import javafx.scene.text.Font;
import javafx.scene.text.FontWeight;
import javafx.scene.text.Text;


/**
 * Displays a progress bar with a text indicator inside
 */
public final class ProgressIndicatorBar extends StackPane {

  private final double totalWork;
  @Nonnull
  private final DoubleProperty workDone = new SimpleDoubleProperty();
  @Nonnull
  private final String labelFormatSpecifier;
  @Nonnull
  private final String unitLabel;

  @Nonnull
  private final ProgressBar progressBar = new ProgressBar();
  @Nonnull
  private final Text text = new Text();

  private static final int DEFAULT_LABEL_PADDING = 2;

  public ProgressIndicatorBar(double maxValue, @Nonnull String labelFormatSpecifier, @Nonnull String unitLabel) {
    this.totalWork = maxValue;
    this.labelFormatSpecifier = labelFormatSpecifier;
    this.unitLabel = unitLabel;
    progressBar.setMaxWidth(Double.MAX_VALUE);
    text.setFill(Color.BLACK);
    text.setFont(Font.font(text.getFont().getFamily(), FontWeight.NORMAL, 14.0));

    syncProgress();

    workDone.addListener((observable, oldValue, newValue) -> syncProgress());

    getChildren().setAll(progressBar, text);
  }

  public double getWorkDone() {
    return workDone.get();
  }

  @Nonnull
  public DoubleProperty workDoneProperty() {
    return workDone;
  }

  // synchronizes the progress indicated with the work done.
  private void syncProgress() {
    if (totalWork < 0) {
      text.setText("");
      progressBar.setProgress(ProgressIndicator.INDETERMINATE_PROGRESS);
    } else {
      text.setText(String.format(labelFormatSpecifier, workDone.get()) + " " + unitLabel);
      progressBar.setProgress(workDone.get() / totalWork);
    }

    progressBar.setMinHeight(text.getBoundsInLocal().getHeight() + DEFAULT_LABEL_PADDING * 2);
    progressBar.setMinWidth (text.getBoundsInLocal().getWidth()  + DEFAULT_LABEL_PADDING * 2);
  }
}

File: commons/javafx/src/main/kotlin/it/neckar/open/javafx/ProgressIndicatorBar.kt
package it.neckar.open.javafx

import it.neckar.open.javafx.properties.*
import javafx.beans.property.DoubleProperty
import javafx.beans.property.SimpleDoubleProperty
import javafx.beans.value.ObservableValue
import javafx.scene.control.ProgressBar
import javafx.scene.control.ProgressIndicator
import javafx.scene.layout.StackPane
import javafx.scene.paint.Color
import javafx.scene.text.Font
import javafx.scene.text.FontWeight
import javafx.scene.text.Text

/**
 * Displays a progress bar with a text indicator inside
 */
class ProgressIndicatorBar(
  val totalWork: Double,
  val labelFormatSpecifier: String,
  val unitLabel: String,
) : StackPane() {

  val workDoneProperty: DoubleProperty = SimpleDoubleProperty()
  var workDone: Double by workDoneProperty

  private val progressBar = ProgressBar()

  private val text = Text()

  init {
    progressBar.maxWidth = Double.MAX_VALUE
    text.fill = Color.BLACK
    text.font = Font.font(text.font.family, FontWeight.BOLD, 14.0)
    syncProgress()
    workDoneProperty.addListener { observable: ObservableValue<out Number?>?, oldValue: Number?, newValue: Number? -> syncProgress() }
    children.setAll(progressBar, text)
  }

  // synchronizes the progress indicated with the work done.
  private fun syncProgress() {
    if (totalWork < 0) {
      text.text = ""
      progressBar.progress = ProgressIndicator.INDETERMINATE_PROGRESS
    } else {
      text.text = String.format(labelFormatSpecifier, workDone) + " " + unitLabel
      progressBar.progress = workDone / totalWork
    }
    progressBar.minHeight = text.boundsInLocal.height + DEFAULT_LABEL_PADDING * 2
    progressBar.minWidth = text.boundsInLocal.width + DEFAULT_LABEL_PADDING * 2
  }

  companion object {
    private const val DEFAULT_LABEL_PADDING = 2
  }
}
